Deriving strategies

SanchayanMaity opened the pull request:
Resolves #43

SanchayanMaity committed:
Add -Wmissing-deriving-strategies to cabal ghc options

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

SanchayanMaity committed:
gitignore: Ignore stack yaml lock files

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

vrom911 commented on membrain.cabal:
```diff
@@ -50,6 +50,7 @@ common common-options
                        -fhide-source-paths
                        -Wmissing-export-lists
                        -Wpartial-fields
+                       -Wmissing-deriving-strategies
```
Could you please put it under the conditional, so it won't produce warning about unknown pragma for earlier GHCs. Like it's done here:
https://github.com/kowainik/tomland/blob/295c2b7580534303723206ba651557adff385cf4/tomland.cabal#L58-L60

SanchayanMaity replied:
Thank you. Pushed a commit fixing this.

vrom911 commented on membrain.cabal:
```diff
@@ -50,6 +50,8 @@ common common-options
                        -Wmissing-export-lists
                        -Wpartial-fields
+if impl(ghc >= 8.8.1)
+  ghc-options:         -Wmissing-deriving-strategies
```
Oh, I know why CI is failing. I see that indentation went incorrect here, so this conditional is applied to the rest of the stanza instead. You should move it 2 spaces right:

```
  if impl(ghc >= 8.8.1)
    ghc-options:       -Wmissing-deriving-strategies
```

SanchayanMaity replied:
Yes, I saw that just now. Sorry about that. Fixed it.

SanchayanMaity committed:
Prevent warning about unknown pragma for earlier GHCs

Put -Wmissing-deriving_strategies under the conditional, so it won't produce
warning about unknown pragma for earlier GHCs.

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

chshersh commented on membrain.cabal:
```diff
@@ -50,6 +50,8 @@ common common-options
                        -Wmissing-export-lists
                        -Wpartial-fields
+  if impl(ghc >= 8.8.1)
+    ghc-options:       -Wmissing-deriving-strategies
```
@SanchayanMaity while we are on this, could you also please error on missing deriving strategies to make sure they are used? You can see how it's done here in `tomland`:

* https://github.com/kowainik/tomland/blob/ff5506eaa3c1a207f1514a56d12ef377b8cdfaa0/tomland.cabal#L60

SanchayanMaity replied:
Pushed. Thank you.

SanchayanMaity committed:
Error on missing deriving strategies

To make sure they are used error on missing deriving strategies.

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

chshersh reviewed (approved):
Thank you for this contribution! :heart:

vrom911 reviewed (approved):
Great stuff! Thank you :+1:
